circuit/src/commonMain/kotlin/com/slack/circuit/Overlay.kt

```diff
@@ -0,0 +1,185 @@
+   * overlay should _always_ signal a result (even if it's just something like `Result.Dismissed`)
+   * so that the [OverlayHost] can properly clear its [currentOverlayData].
+   *
+   * If no data is needed in a result, use [Unit] for the result type.
+   *
+   * This function should only be called from UI contexts and _not_ presenters, as overlays are a UI
+   * concern.
+   */
+  public suspend fun <Result : Any> show(overlay: Overlay<Result>): Result
+}
+
+/** Returns a remembered an [OverlayHost] that can be used to show overlays. */
+@Composable public fun rememberOverlayHost(): OverlayHost = remember { OverlayHostImpl() }
+
+private class OverlayHostImpl : OverlayHost {
+  /**
+   * Only one [Overlay] can be shown at a time. Since a suspending Mutex is a fair queue, this
+   * manages our message queue and we don't have to maintain one.
+   */
+  private val mutex = Mutex()
+
+  override var currentOverlayData: OverlayHostData<Any>? by mutableStateOf(null)
+    private set
+
+  override suspend fun <T : Any> show(overlay: Overlay<T>): T =
+    mutex.withLock {
+      try {
+        return suspendCancellableCoroutine { continuation ->
+          @Suppress("UNCHECKED_CAST")
+          currentOverlayData =
+            OverlayHostDataImpl(
+              overlay as Overlay<Any>,
+              continuation as CancellableContinuation<Any>
+            )
+        }
+      } finally {
+        currentOverlayData = null
+      }
+    }
+}
+
+/**
+ * Data managed by an [OverlayHost] to track the current overlay state. This should rarely be
+ * implemented by consumers!
+ */
+@Stable
+public interface OverlayHostData<Result : Any> {
+  /** The [Overlay] that is currently being shown. Read-only. */
+  public val overlay: Overlay<Result>
+
+  /**
+   * Invoked to finish the current overlay with the given [result]. This should be called by
+   * wherever the [OverlayHost] is being managed.
+   */
+  public fun finish(result: Result)
+}
+
+private class OverlayHostDataImpl<T : Any>(
+  override val overlay: Overlay<T>,
+  private val continuation: CancellableContinuation<T>,
+) : OverlayHostData<T> {
+  override fun finish(result: T) {
+    if (continuation.isActive) continuation.resume(result)
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (javaClass != other?.javaClass) return false
+
+    other as OverlayHostDataImpl<*>
+
+    if (overlay != other.overlay) return false
+    if (continuation != other.continuation) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    var result = overlay.hashCode()
+    result = 31 * result + continuation.hashCode()
+    return result
+  }
+}
+
+/**
+ * An [OverlayNavigator] is a simple API offered to [overlays][Overlay] to call [finish] with a
+ * result when they are done.
+ */
+@Stable
+public fun interface OverlayNavigator<Result : Any> {
+  /** Called by the [Overlay] with a [result] it's done. */
+  public fun finish(result: Result)
```

Comment:
I think I might not have communicated that thought well, I only meant I thought it wasn't a good use case because it was a bit surprising that starring it would dismiss a detail screen (versus just pushing it into the data layer). The detail screen doesn't feel like ephemeral UI, nor did starring seem like the CTA/primary action on it that would incur a dismissal.

I think the other two bits are cases we'd usually expect to push into a data layer on DI or a composite presenter, but I think I just need to see some examples. It would be weird to me for actions taken on a bottom sheet or dialog to update the UI behind it without dismissing the modal 🤔

Reply:
> It would be weird to me for actions taken on a bottom sheet or dialog to update the UI behind it without dismissing the modal 

I agree it wouldn't make much sense for a dialog, but this feels like a common pattern for a bottom sheet. It's also something we currently do in slack-android (when attaching files to a draft for example). Anyway, this approach just feels very limiting, and I'd rather be flexible than prescriptive. Especially when there is no technical reason (that I'm aware of) for doing so.

I think we should merge this and continue the discussion, so nothing to do at the moment

Reply:
Definitely would need to see an example, I can't think of any. The drafts example (on iOS at least) doesn't update the content behind the sheet until you dismiss it 🤔. I don't think this is limiting, just one solution for a common case (where the others are hoisted state in DI or composite presenters).

Merging on the basis of the above though! Can continue discussion later.